### PR TITLE
Migrate all models to make full use of the Module base class

### PR DESCRIPTION
**Pull request by friedrich**
### Ticket

https://github.com/tenstorrent/tt-metal/issues/36034

### What's changed

All tt_dit models are now derived from the `Module` base class. Custom PyTorch state loading, as well as custom cache loading and saving was replaced by the corresponding functionality of `Module`. Instantiating modules from PyTorch modules was simplified. A alternative set of model caching helpers was introduced to make custom cache handling more straightforward.

The following models were migrated:

- CLIP encoder
- T5 encoder
- WAN DiT
- Mochi DiT
- Stable Diffusion 3.5 DiT
- WAN VAE encoder
- WAN VAE decoder
- Mochi VAE decoder
- Stable Diffusion 3.5 VAE decoder

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:friedrich/caching)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:friedrich/caching)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:friedrich/caching)
- [ ] New/Existing tests provide coverage for changes

#### Model tests

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-multi-card-demo-tests.yaml?query=branch:friedrich/caching)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-integration-tests.yaml?query=branch:friedrich/caching) : Motif hanging on main. Runner issues for Flux and Wan. Pass locally and [previously](https://github.com/tenstorrent/tt-metal/actions/runs/21751106905).
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml?query=branch:friedrich/caching): Some tt_dit jobs cancelled, but [passed before rebase](https://github.com/tenstorrent/tt-metal/actions/runs/21721096459/job/62651771496). Uncanceled tt_dit jobs succeded.
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml?query=branch:friedrich/caching) : Mochi, SD35 and Qwen Image experience runner issues. Passes locally.
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-integration-tests.yaml?query=branch:friedrich/caching): Mochi also failing on [main](https://github.com/tenstorrent/tt-metal/actions/runs/21846081187/job/63047615617) for the same reason (exit code 137)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml?query=branch:friedrich/caching): Non tt_dit failures unrelated to the changes.
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-perf-tests.yaml?query=branch:friedrich/caching): Same perf regressions as [main](https://github.com/tenstorrent/tt-metal/actions/runs/21864778373).

### CODE REVIEW

File: models/tt_dit/layers/module.py [hunk 1]
```diff
@@ -154,27 +154,24 @@ def load_torch_state_dict(self, state_dict: Mapping[str, torch.Tensor], *, stric
...
 
-    def save(self, directory: str | Path, /) -> None:
+    def save(self, directory: str | Path, /, *, prefix: str = "") -> None:
         directory = Path(directory)
-        directory.mkdir(exist_ok=True)
+        directory.mkdir(exist_ok=True, parents=True)
 
         for name, child in self.named_children():
-            child.save(directory / name)
+            child.save(directory, prefix=f"{prefix}{name}.")
```
[the remark this answers is not in the document]
Comment (friedrich) [reply models/tt_dit/layers/module.py:163]: I think this comment ist just wrong.

File: models/tt_dit/utils/conv3d.py [hunk 2]
```diff
@@ -154,7 +113,7 @@ def aligned_channels(channels):
     return channels
 
 
-def prepare_conv3d_weights(mesh_device, weight, bias, conv_config, ALIGNMENT=ALIGNMENT):
+def prepare_conv3d_weights(weight, bias, conv_config):
     """Prepare weights and bias for TTNN."""
```
[the remark this answers is not in the document]
Comment (friedrich) [reply models/tt_dit/utils/conv3d.py:119]: The docstring is correct.

File: models/tt_dit/pipelines/mochi/pipeline_mochi.py [hunk 3]
```diff
@@ -166,6 +166,13 @@ def __init__(
         self.num_links = num_links
         self.reload_dit_model = self.mesh_device.get_num_devices() <= 8  # Only required if VAE is memory-constrained.
 
+        if self.reload_dit_model and not cache.cache_dir_is_set():
+            msg = (
+                "Cache must be enabled when DiT model reloading is enabled (reload_dit_model=True)."
+                "Please set TT_DIT_CACHE_DIR environment variable to enable caching."
```
[the remark this answers is not in the document]
Comment (friedrich) [reply models/tt_dit/pipelines/mochi/pipeline_mochi.py:170]: Fixed.

File: models/experimental/tt_dit/tests/models/sd35/test_transformer_sd35.py [hunk 4]
```diff
@@ -266,19 +266,20 @@ def test_sd35_transformer2d_model(
...
+        try:
+            cache.load_model(
+                tt_model,
+                model_name="stable-diffusion-3.5-large",
+                subfolder="transformer",
+                parallel_config=parallel_config,
+                mesh_shape=tuple(mesh_device.shape),
+            )
+        except cache.MissingCacheError as err:
+            msg = "Cache path does not exist. Run test_mochi_transformer_model_caching first with the desired parallel config."
```
[the remark this answers is not in the document]
Comment (friedrich) [reply models/experimental/tt_dit/tests/models/sd35/test_transformer_sd35.py]: Fixed.

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 5]
```diff
@@ -342,16 +348,14 @@ def _load_transformer2(self):
...
-            self.parallel_config,
-            tuple(self.mesh_device.shape),
-        ):
-            logger.info("Loading transformer weights from PyTorch state dict")
-            self.transformer_2.load_torch_state_dict(self.torch_transformer_2.state_dict())
+            model_name=os.path.basename(self.checkpoint_name),
+            subfolder="transformer_2",
+            parallel_config=self.parallel_config,
+            mesh_shape=tuple(self.mesh_device.shape),
+            get_torch_state_dict=lambda: self.torch_transformer_2.state_dict(),
```
Comment (sadesoyeTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py:357]: Any reason to not just pass the dict as opposed to a callable?

Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:357]: Yes, `load_model` takes a callable to make it possible to load the PyTorch model only if the cache does not exist, which speeds up loading. This feature is not used yet here, that is why the callable is trivially returning the already loaded state dict.

Comment (sadesoyeTT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:357]: Make sense. Please add a comment for it so folks can take advantage of the feature as needed.

**Review by sadesoyeTT — APPROVED** [review]

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 5 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:357]: https://github.com/tenstorrent/tt-metal/pull/37117/changes/d0cef1357f089aeb60727269606970c20d420d57

**Review by cglagovichTT — APPROVED** [review]
